### PR TITLE
Creating quartz scheduler

### PR DESCRIPTION
**Pull request by kunkun-tang**
Inspired by #1488 , this PR proposes bringing quartz to AZ. Quartz Scheduler provides 3 modes of job stores: RAMJobStore, JDBCJobStore, and TerraCottaJobStore. In order to persist the job statuses out of Application memory regardless of downtime, we recommend using `JDBC JobStore`. In fact, users can specify the mode in quartz properties.

This code patch creates Quartz server and manages Quartz API (e.g., start, pause, shut down) at class QuartzScheduler. We directly use h2-in-memory DB to directly test JDBC Quartz in unit test.

### CODE REVIEW

**Comment by kunkun-tang** [comment]
@chengren311 You might want to take a look at correponding tests in this PR. Let me know if you have any questions.

**Comment by HappyRay** [comment]
Could you add more details to the change description? Duplication between the change description and the issue is ok IMHO. It makes it easier for readers.

File: az-core/src/main/java/azkaban/Constants.java [hunk 1]
```diff
@@ -151,6 +151,9 @@
      *    implies save latest 3 versions saved in storage.
      **/
     public static final String AZKABAN_STORAGE_ARTIFACT_MAX_RETENTION = "azkaban.storage.artifact.max.retention";
+
+    // enable Quartz Scheduler if true.
+    public static final String ENABLE_QUARTZ= "enable.quartz";
```
Comment (HappyRay) [inline az-core/src/main/java/azkaban/Constants.java]: prefix it with azkaban.
Maybe
azkaban.server.schedule.enable_quartz?

File: azkaban-db/src/main/sql/create.quartz-tables-all.sql [hunk 2]
```diff
@@ -0,0 +1,165 @@
+-- This file collects all quartz table create statement required for quartz 2.2.1
```
Comment (HappyRay) [inline azkaban-db/src/main/sql/create.quartz-tables-all.sql]: The current convention is to create one .sql file per table. 
And the database setup code relies on it.

Is there any reason to break this convention?

File: azkaban-web-server/src/test/java/azkaban/scheduler/TestQuartzJob.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
+/*
```
Comment (HappyRay) [inline azkaban-web-server/src/test/java/azkaban/scheduler/TestQuartzJob.java]: Other tests are named with xxxTest.java.

File: azkaban-web-server/src/test/resources/create.quartz-all-tables.sql [hunk 4]
```diff
@@ -0,0 +1,165 @@
+-- All necessary quartz table create statement collection for unit test purpose
```
Comment (HappyRay) [inline azkaban-web-server/src/test/resources/create.quartz-all-tables.sql]: If the sql statements are changed incorrectly and this file is not updated, the tests may pass.

Can you think of ways to remove this duplication?

File: az-core/src/main/java/azkaban/Constants.java (hunk 1 above)
Comment (kunkun-tang) [reply az-core/src/main/java/azkaban/Constants.java]: why do we need a long name? I like `azkaban.enable_quartz`?

File: azkaban-db/src/main/sql/create.quartz-tables-all.sql (hunk 2 above)
Comment (kunkun-tang) [reply azkaban-db/src/main/sql/create.quartz-tables-all.sql]: these tables originally reside in one file together. I would believe we should not manage/modify these sql statements after those tables are built, as we should not change quartz code, right? In this way, why not just have a big file, which just serves as a reference to users.

File: azkaban-web-server/src/test/java/azkaban/scheduler/TestQuartzJob.java (hunk 3 above)
Comment (kunkun-tang) [reply azkaban-web-server/src/test/java/azkaban/scheduler/TestQuartzJob.java]: This is a not test, right? Is SampleQuartzJob a good name?

File: azkaban-web-server/src/test/resources/create.quartz-all-tables.sql (hunk 4 above)
Comment (kunkun-tang) [reply azkaban-web-server/src/test/resources/create.quartz-all-tables.sql]: I can not think of situations where we will change quartz SQL statements. Do I miss something?

File: azkaban-web-server/src/main/java/azkaban/scheduler/AbstractQuartzJob.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+package azkaban.scheduler;
+
+import java.io.Serializable;
+import org.quartz.Job;
+import org.quartz.JobExecutionContext;
+
+public abstract class AbstractQuartzJob implements Job {
+
+  protected static <T extends Serializable> T asT(final Object service, final Class<T> type) {
```
Comment (chengren311) [inline azkaban-web-server/src/main/java/azkaban/scheduler/AbstractQuartzJob.java]: please document these methods

File: azkaban-web-server/src/main/java/azkaban/scheduler/QuartzJobDescription.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  public QuartzJobDescription(final Class<T> jobClass,
+      final String groupName,
+      final Map<String, ? extends Serializable> contextMap) {
+
+    /**
+     * This check is necessary for raw type. Please see test
+     * {@link QuartzJobDescriptionTest#testCreateQuartzJobDescription2}
+     */
+    if (jobClass.getSuperclass() != AbstractQuartzJob.class) {
+      throw new RuntimeException("jobClass must extend AbstractQuartzJob class");
```
Comment (chengren311) [inline azkaban-web-server/src/main/java/azkaban/scheduler/QuartzJobDescription.java]: i would prefer a more meaningful exception type, maybe like illegalargumentexception?

File: az-core/src/main/java/azkaban/Constants.java (hunk 1 above)
Comment (HappyRay) [reply az-core/src/main/java/azkaban/Constants.java]: I am less worried about the length. We can copy and paste easily.
IMHO a more descriptive name is better. We would like to maintain a clear and consistent name hierarchy and namespace going forward.

File: azkaban-db/src/main/sql/create.quartz-tables-all.sql (hunk 2 above)
Comment (HappyRay) [reply azkaban-db/src/main/sql/create.quartz-tables-all.sql]: I see. 
So you are ok with breaking the database setup code then?
Could you come up with a plan for how to evolve DB schemas?

File: azkaban-web-server/src/test/java/azkaban/scheduler/TestQuartzJob.java (hunk 3 above)
Comment (HappyRay) [reply azkaban-web-server/src/test/java/azkaban/scheduler/TestQuartzJob.java]: I see. Then the name is fine.

File: azkaban-web-server/src/test/resources/create.quartz-all-tables.sql (hunk 4 above)
Comment (HappyRay) [reply azkaban-web-server/src/test/resources/create.quartz-all-tables.sql]: When we update Quartz, the sql statements may change. Am I right?

File: azkaban-web-server/src/main/java/azkaban/scheduler/QuartzJobDescription.java (hunk 6 above)
Comment (kunkun-tang) [reply azkaban-web-server/src/main/java/azkaban/scheduler/QuartzJobDescription.java]: I found that CastException seems more appropriate. Updated

File: azkaban-web-server/src/main/java/azkaban/scheduler/AbstractQuartzJob.java (hunk 5 above)
Comment (kunkun-tang) [reply azkaban-web-server/src/main/java/azkaban/scheduler/AbstractQuartzJob.java]: Wrote some java doc on top of it.

File: az-core/src/main/java/azkaban/Constants.java (hunk 1 above)
Comment (kunkun-tang) [reply az-core/src/main/java/azkaban/Constants.java]: Updated.

File: azkaban-db/src/main/sql/create.quartz-tables-all.sql (hunk 2 above)
Comment (kunkun-tang) [reply azkaban-db/src/main/sql/create.quartz-tables-all.sql]: What do you mean by breaking database setup code? I believe new DatabaseSetup should be able to deal with both the two cases: 1. Sql folders; 2. one Sql file.

File: azkaban-web-server/src/test/resources/create.quartz-all-tables.sql (hunk 4 above)
Comment (kunkun-tang) [reply azkaban-web-server/src/test/resources/create.quartz-all-tables.sql]: Upgrade Quartz! Yes, that's quite possible.

File: azkaban-web-server/src/main/java/azkaban/scheduler/QuartzScheduler.java [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    final StdSchedulerFactory schedulerFactory =
+        new StdSchedulerFactory(azProps.toProperties());
+    this.scheduler = schedulerFactory.getScheduler();
+  }
+
+  public void start() {
+    try {
+      this.scheduler.start();
+    } catch (final SchedulerException e) {
+      logger.warn("Error starting Quartz scheduler: ", e);
```
Comment (chengren311) [inline azkaban-web-server/src/main/java/azkaban/scheduler/QuartzScheduler.java]: warn seems too relaxing for this kind of exception. maybe error?

File: azkaban-web-server/src/main/java/azkaban/scheduler/QuartzScheduler.java [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    } catch (final SchedulerException e) {
+      logger.error("Exception shutting down scheduler: ", e);
+    }
+  }
+
+  public void unregisterJob(final String groupName) throws SchedulerException {
+    if(!ifJobExist(groupName)) {
+      logger.warn("can not find job with " + groupName + " in quartz.");
+    } else {
+      this.scheduler.deleteJob(new JobKey("job1", groupName));
```
Comment (chengren311) [inline azkaban-web-server/src/main/java/azkaban/scheduler/QuartzScheduler.java]: why job1?

File: azkaban-web-server/src/main/java/azkaban/scheduler/QuartzScheduler.java [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+        .build();
+
+    this.scheduler.scheduleJob(job, trigger);
+    logger.info("Quartz Schedule with jobDetail " + job.getDescription() + " is registered.");
+  }
+
+
+  public boolean ifJobExist(final String groupName) throws SchedulerException {
+    final Set<JobKey> jobKeySet = this.scheduler.getJobKeys(GroupMatcher.jobGroupEquals(groupName));
+    return jobKeySet !=null && jobKeySet.size() > 0;
```
Comment (chengren311) [inline azkaban-web-server/src/main/java/azkaban/scheduler/QuartzScheduler.java]: space before null

File: azkaban-web-server/src/main/java/azkaban/scheduler/QuartzScheduler.java [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+  }
+
+  Scheduler getScheduler() {
```
Comment (chengren311) [inline azkaban-web-server/src/main/java/azkaban/scheduler/QuartzScheduler.java]: access modifiers?

File: azkaban-web-server/src/main/java/azkaban/scheduler/AbstractQuartzJob.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+package azkaban.scheduler;
+
+import java.io.Serializable;
+import org.quartz.Job;
+import org.quartz.JobExecutionContext;
+
+public abstract class AbstractQuartzJob implements Job {
+
+  /**
+   * Cast the object to the original one with the type. The object must extend Serializable.
```
Comment (chengren311) [inline azkaban-web-server/src/main/java/azkaban/scheduler/AbstractQuartzJob.java]: It is still not clear to me the purpose of the method with the description.

File: azkaban-web-server/src/main/java/azkaban/scheduler/QuartzScheduler.java [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    if (!CronExpression.isValidExpression(cronExpression)) {
+      throw new SchedulerException("The cron expression string <" +  cronExpression + "> is not valid.");
+    }
+
+    // TODO kunkun-tang: Today all JobDetail should use "job1", and we will modify this when we
+    // start supporting multi schedules per flow.
+    final JobDetail job = JobBuilder.newJob(jobDescription.getJobClass())
+        .withIdentity("job1", jobDescription.getGroupName()).build();
```
Comment (chengren311) [inline azkaban-web-server/src/main/java/azkaban/scheduler/QuartzScheduler.java]: then can we make job1 a constant?

File: azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServer.java [hunk 13]
```diff
@@ -184,8 +189,13 @@ public AzkabanWebServer(final Props props,
       DateTimeZone.setDefault(DateTimeZone.forID(timezone));
       logger.info("Setting timezone to " + timezone);
     }
-
     configureMBeanServer();
+
+    // Start Quartz Server
```
Comment (chengren311) [inline azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServer.java]: the comment is unnecessary.

File: azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServer.java [hunk 14]
```diff
@@ -184,8 +189,13 @@ public AzkabanWebServer(final Props props,
...
+    if(props.containsKey(ConfigurationKeys.ENABLE_QUARTZ) && props.getBoolean(ConfigurationKeys
+        .ENABLE_QUARTZ)) {
+      quartzScheduler.start();
```
Comment (chengren311) [inline azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServer.java]: should we start the scheduler in constructor or prepareAndStartServer()?

File: azkaban-web-server/src/main/java/azkaban/scheduler/QuartzScheduler.java (hunk 12 above)
Comment (kunkun-tang) [reply azkaban-web-server/src/main/java/azkaban/scheduler/QuartzScheduler.java]: Good suggestion. Adding

File: azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServer.java (hunk 14 above)
Comment (kunkun-tang) [reply azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServer.java]: What's the benefits over that approach?

File: azkaban-web-server/src/main/java/azkaban/scheduler/QuartzScheduler.java (hunk 10 above)
Comment (kunkun-tang) [reply azkaban-web-server/src/main/java/azkaban/scheduler/QuartzScheduler.java]: Adding public modifier here. I will add a shut down hook with web server stop.

File: azkaban-web-server/src/main/java/azkaban/scheduler/QuartzScheduler.java (hunk 9 above)
Comment (kunkun-tang) [reply azkaban-web-server/src/main/java/azkaban/scheduler/QuartzScheduler.java]: added.

File: azkaban-web-server/src/main/java/azkaban/scheduler/QuartzScheduler.java (hunk 8 above)
Comment (kunkun-tang) [reply azkaban-web-server/src/main/java/azkaban/scheduler/QuartzScheduler.java]: explained in a comment.

File: azkaban-web-server/src/main/java/azkaban/scheduler/QuartzScheduler.java (hunk 10 above)
Comment (kunkun-tang) [reply azkaban-web-server/src/main/java/azkaban/scheduler/QuartzScheduler.java]: Seems that shutdown doesn't need to `getScheduler`. For now, this method is never called by anyone.

File: azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServer.java (hunk 14 above)
Comment (chengren311) [reply azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServer.java]: to separate preparation work from constructor.

Comment (kunkun-tang) [reply azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServer.java]: Moving to prepareAndStartServer()

**Pull request merged**